lib/chef/key.rb

```diff
@@ -0,0 +1,195 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+require 'chef/json_compat'
+require 'chef/mixin/params_validate'
+
+class Chef
+  class Key
+
+    include Chef::Mixin::ParamsValidate
+
+    attr_reader :actor_field_name
+
+    def initialize(actor, actor_field_name)
+      # Actor that the key is for, either a client or a user.
+      @actor = actor
+
+      unless actor_field_name == "user" || actor_field_name == "client"
+        raise ArgumentError.new("the second argument to initialize must be either 'user' or 'client'")
+      end
+
+      @actor_field_name = actor_field_name
+
+      @name = nil
+      @public_key = nil
+      @expiration_date = nil
+    end
+
+    def actor(arg=nil)
+      set_or_return(:actor, arg,
+                    :regex => /^[a-z0-9\-_]+$/)
+    end
+
+    def name(arg=nil)
+      set_or_return(:name, arg,
+                    :regex => /^[a-z0-9\-_]+$/)
+    end
+
+    def public_key(arg=nil)
+      set_or_return(:public_key, arg,
+                    :kind_of => String)
+    end
+
+    def expiration_date(arg=nil)
+      set_or_return(:expiration_date, arg,
+                    :regex => /^(\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}Z|infinity)$/)
+    end
+
+    def to_hash
+      result = {
+        @actor_field_name => @actor,
+      }
+      result["name"] = @name if @name
+      result["public_key"] = @public_key if @public_key
+      result["expiration_date"] = @expiration_date if @expiration_date
+      result
+    end
+
+    def to_json(*a)
+      Chef::JSONCompat.to_json(to_hash, *a)
+    end
+
+    def create
+      if @public_key.nil? || @expiration_date.nil?
+        raise ArgumentError.new("public_key and expiration_date fields must be populated when create is called")
+      end
+
+      # defaults the key name to the fingerprint of the key
+      if @name == nil
+        # TODO: is it safe to assume we aren't dealing with certs here?
+        openssl_key_object = OpenSSL::PKey::RSA.new(@public_key)
+        data_string = OpenSSL::ASN1::Sequence([
+                                                OpenSSL::ASN1::Integer.new(openssl_key_object.public_key.n),
+                                                OpenSSL::ASN1::Integer.new(openssl_key_object.public_key.e)
+                                              ])
+        @name = OpenSSL::Digest::SHA1.hexdigest(data_string.to_der).scan(/../).join(':')
+      end
+
+      payload = {"name" => @name, "public_key" => @public_key, "expiration_date" => @expiration_date}
+      if @actor_field_name == "user"
```

Comment:
Rather than switching like this in all of the instance methods, perhaps something like:

```ruby
def chef_rest
  @rest ||= if @actor_field_name == "user
                   Chef::REST.new(Chef::Config[:chef_server_root])
                else
                  Chef::REST.new(Chef::Config[:chef_server_url])
              end
end

def api_base
  @api_base ||= if @actor_field_name == "user
    "users"
  else
   "clients"
  end
end
```

then, in your methods you can just do:

```
chef_rest.post_rest("#{api_base}/#{@actor}/keys")
```

with no switching.